src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java

```diff
@@ -730,6 +730,29 @@ protected AuthorizationCodeFlow buildAuthorizationCodeFlow() {
         return builder.build();
     }
 
+    private String getValidRedirectUrl(String url) {
+        if (url != null && !url.isEmpty()) {
+            // Check if the URL is relative and starts with a slash
+            if (url.startsWith("/")) {
+                return getRootUrl() + url; // Convert to absolute URL
+            }
+            // If not relative, then check if it's a valid absolute URL
+            try {
+                URL parsedUrl = new URL(url);
+                String host = parsedUrl.getHost();
+                String expectedHost = new URL(getRootUrl()).getHost();
+                // Check if the host matches the Jenkins domain
+                if (host.equals(expectedHost)) {
+                    return url; // The URL is absolute and valid
+                }
+            } catch (MalformedURLException e) {
+                // Invalid absolute URL, will return root URL
+            }
+        }
+        // If the URL is null, empty, or invalid, return the root URL
+        return getRootUrl();
```

Comment:
Root URL is not always correctly setup (typicallty localhost on k8s deployments) and people rely on referer to deduce the correct UTL as seen by the user.

Reply:
See my comment in the issue, please. The referer is passed in when calling the function.